src/main/java/org/assertj/core/api/ComparatorBasedObjectAssert.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2020 the original author or authors.
+ */
+package org.assertj.core.api;
+
+import java.util.Comparator;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.assertj.core.description.Description;
+import org.assertj.core.presentation.Representation;
+
+/**
+ * @param <ACTUAL> 
+ *
+ * @author Stefano Cordio
+ * @since 3.19.0
+ */
+public class ComparatorBasedObjectAssert<ACTUAL> extends ObjectAssert<ACTUAL>
+    implements ComparisonBasedAssert<ComparatorBasedObjectAssert<ACTUAL>, ACTUAL> {
+
+  public ComparatorBasedObjectAssert(ACTUAL actual, Comparator<? super ACTUAL> comparator, String comparatorDescription) {
+    super(actual);
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isEqualByComparingTo(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isNotEqualByComparingTo(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isLessThan(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isLessThanOrEqualTo(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isGreaterThan(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isGreaterThanOrEqualTo(ACTUAL other) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isBetween(ACTUAL startInclusive, ACTUAL endInclusive) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> isStrictlyBetween(ACTUAL startExclusive, ACTUAL endExclusive) {
+    // FIXME assertion
+    return this;
+  }
+
+  @Override
+  public ComparatorBasedObjectAssert<ACTUAL> as(Description description) {
+    return (ComparatorBasedObjectAssert<ACTUAL>) super.as(description);
+  }
```

Comment:
not sure it it is a better approach but we would not need to cast if we were returning `myself`:
```
  @Override
  @CheckReturnValue
  public SELF as(Description description) {
    return super.as(description);
  }
```
Not 100% sure if we need to annotate with `@CheckReturnValue` as the overridden method has already the annotation

Reply:
`myself` is tricky in this case. As we inherit from `ObjectAssert` and not from `AbstractObjectAssert`, `myself` has `ObjectAssert<ACTUAL>` type. We could redefine another `myself` field in `ComparatorBasedObjectAssert` or just use `this` and then the pattern would become something like:

```java
  @Override
  public ComparatorBasedObjectAssert<ACTUAL> as(Description description) {
    super.as(description);
    return this;
  }
```

I preferred the casting not to miss any potential manipulation of myself in any of the superclasses, like methods doing:
```java
    return myself.with...();
```
but I'm not fully sure if it is needed as this might happen only for methods returning a new assertion object and these should not be overridden. I need to double-check this.

Inheriting from `AbstractObjectAssert` was not an option because I needed to specialize the return type of `usingComparator` methods, and I think that as long as we have SELF the type invariance of generics is preventing the specialization. That's why I fall back to `ObjectAssert` where we have the concrete type, which can be specialized to a subtype. I'm open to any better approach!